### PR TITLE
Add Auto Vacuum Settings to PostgreSQL

### PR DESCRIPTION
**Pull request by npwalker**
In the past I've seen the console and puppetdb databases grow
much larger than expected and I have a hypothesis that this is
because the default 20% setting for autovacuum is too high for our
workload.  So, there's a potential that changing this to 1%
might be too low and might reduce performance but it should
reduce disk usage and it can always be tuned upward if the
performance hit is too large, however, it is much more difficult
to reduce the size of a database after it has grown too large.

Basically, I'm erring on the side of using less disk space and
hopefully less outages due to running out of disk space and we
can open a conversation on performance later if one needs to be had.

### CODE REVIEW

**Comment by glarizza** [comment]
Have you started using this with customers already?  What's been the feedback?

**Comment by npwalker** [comment]
I've set the autovacuum settings on specific large tables success.  Tables like resource_statuses in the console would become extremely bloated but setting these settings to 1% would keep the table size in check with no reported performance issues.  

The difference here is that I'm changing the autovacuum setting on all tables in all databases.  So, autovacuum will more regularly do work which I think is what we want.  Otherwise autovacuum runs infrequently and causes longer duration spikes in usage ( and potentially doesn't fully finish the vacuum process ).  

I assume the risk in setting this low is for large tables that need lots of vacuuming so also enabling this for smaller tables seems like a non-issue but as I say in the comment I'm willing to trade a hypothetical performance hit for consistent disk size for the database.  As slow performance is something we can always improve with tuning but running out of disk space turns out to be an urgent issue we can avoid.

**Comment by trlinkin** [comment]
Shouldn't the ability to tune something PE related like this go in the `puppet_enterprise` module and be part of the product from the get-go?

**Comment by glarizza** [comment]
Chicken and egg - it should be in `puppet_enterprise` if it's been proven to be successful and stable, but we can't prove that till we use it.  I consider people using `control-repo` and the provided profiles to be customers who are willing to tune their infrastructure and stay on a PS workflow, so I'd accept that there's more flexibility here than in `puppet_enterprise`.  My 2 cents.

**Comment by trlinkin** [comment]
Yeah, ok. I wasn't suggesting that it be set to 1% in `puppet_enterprise` but merely that it be possible and then we could easily suggest to anyone to tweak that value.

So am I to understand that this control repo has grown beyond being a beginner example and starting point to something that is considered somewhat leading edge thinking for those  daring enough to use it?

**Comment by glarizza** [comment]
Hmm, that's a better question...and I'm not entirely certain how to answer that.  When does this become too big?  Maybe we need a puppetlabs-profiles repo to bring in the profiles at this point to facilitate outside development (and so when @npwalker makes changes to them it doesn't screw everyone's control-repo up).  Whatcha think @npwalker?

**Comment by npwalker** [comment]
I think we've officially surpassed discussing this PR and now we've moved onto the philosophy of this repo.  So, we should probably take up the conversation in a new forum.  :)

**Comment by npwalker** [comment]
To close the loop on the PR question... we don't manage postgresql out of the box in PE so I'm putting it here and I'll ask for a feature to manage the setting but that would require that we manage postgresql post install which seems like a large change.

**Comment by glarizza** [comment]
OKAY, so there was a side discussion and I want to provide that information here:

1. This repo is meant to be a start for general customers, so PSEs use this for that purpose
2. Nick uses this repo with his customers who have specific questions and tuning needs (i.e. the autovacuum settings).  Nick takes advantage of Hiera data and custom facts to do this work, and thus if we broke profiles out into their own repo we'd need to use the experimental data-in-modules bit to bundle/package that up, or do some hacky params work...so there's a benefit to having the profiles here.  There's also benefit in having roles/profiles inside the control repo because of all the reasons we know (namely - the webhook or filesync)
3. The customers PSEs are working with definitely may/will have these questions and needs, but we shouldn't DEFAULT anyone to the specialized settings/tunings that are done here...this repo should, out of the box, just setup base configuration (like Code Manager or R10k, for example)
4. Should ANY customer want to enable the tuning settings, they have the ability to set custom facts or follow the docs in the repo to enable them...we're providing them for visibility, and, in my/our opinion, visibility is definitely better than it being unknown or black magic

So I think we're cool with this PR since the settings aren't on BY DEFAULT - you need to do some custom fact tuning to enable them.  That's how we'll proceed:  any tuning requires customer intervention in SOME way and isn't there by default.  Sound good?  If no opposition I'll merge by EOD today

**Comment by natemccurdy** [comment]
I'm 100% on board with this being a beginning or entry-point into using control repos. Having to hack pe-postgres right off the bat seems like a bit much, and will definitely overcomplicate things for people new to Puppet Enterprise.

That said, this is a useful setting to tune, but I don't think it should be on by default which, correct me if I'm wrong, it is due to the class being included in `site/role/manifests/all_in_one_pe.pp`.

**Comment by fvoges** [comment]
:+1: to have tuning here, but turned off by default

**Comment by npwalker** [comment]
Okay, I moved this code over into the pe_databases module.  

https://github.com/npwalker/pe_databases/blob/master/manifests/postgresql_settings.pp

So the recommendation is essentially to make a new role that includes this class and the extra stuff that may come along in the future.  So, they're be at least two choices.  A basic choice that is just code manager and some of the puppet master settings and then an option that includes more classes and best practices?

**Comment by npwalker** [comment]
Closing this as it is outside the scope of this repo.
